eth/beacon/state_machines/base.py

```diff
@@ -193,9 +222,562 @@ def active_state(self) -> ActiveState:
+            shard_block_hash +
+            justified_slot.to_bytes(8, 'big')
+        )
+
+    #
+    # Import block API
+    #
+    def import_block(
+            self,
+            block: BaseBeaconBlock) -> Tuple[BaseBeaconBlock, CrystallizedState, ActiveState]:
+        """
+        Import the given block to the chain.
+        """
+        processing_block = self.get_block_class().from_block(block)
+        processing_block, processed_crystallized_state, processed_active_state = self.process_block(
+            self.crystallized_state,
+            self.active_state,
+            processing_block,
+            self.chaindb,
+            self.config,
+        )
+
+        # Validate state roots
+        self.validate_state_roots(
+            processed_crystallized_state.hash,
+            processed_active_state.hash,
+            block,
+        )
+
+        self.block = processing_block
+        self._update_the_states(processed_crystallized_state, processed_active_state)
+        # TODO: persist states in BeaconChain if needed
+
+        return self.block, self.crystallized_state, self.active_state
+
+    #
+    # Process block APIs
+    #
+    @classmethod
+    def process_block(
+            cls,
+            crystallized_state: CrystallizedState,
+            active_state: ActiveState,
+            block: BaseBeaconBlock,
+            chaindb: BaseBeaconChainDB,
+            config: BeaconConfig) -> Tuple[BaseBeaconBlock, CrystallizedState, ActiveState]:
+        """
+        Process ``block`` and return the new crystallized state and active state.
+        """
+        # Process per block state changes (ActiveState)
+        processing_active_state = cls.compute_per_block_transtion(
+            crystallized_state,
+            active_state,
+            block,
+            chaindb,
+            config.CYCLE_LENGTH,
+        )
+
+        # Process per cycle state changes (CrystallizedState and ActiveState)
+        processed_crystallized_state, processed_active_state = cls.compute_cycle_transitions(
+            crystallized_state,
+            processing_active_state,
+        )
+
+        # Return the copy
+        result_block = block.copy()
+        return result_block, processed_crystallized_state, processed_active_state
+
+    @classmethod
+    def compute_per_block_transtion(cls,
+                                    crystallized_state: CrystallizedState,
+                                    active_state: ActiveState,
+                                    block: BaseBeaconBlock,
+                                    chaindb: BaseBeaconChainDB,
+                                    cycle_length: int) -> ActiveState:
+        """
+        Process ``block`` and return the new ActiveState.
+
+
+        TODO: It doesn't match the latest spec.
+        There will be more fields need to be updated in ActiveState.
+        """
+        parent_block = chaindb.get_block_by_hash(block.parent_hash)
```

Comment:
@pipermerriam I left some TODOs hints in [`BeaconChainDB`](https://github.com/ethereum/py-evm/blob/96fc4ede33378823b470f6540245b1fc94752d1d/eth/beacon/db/chain.py#L635-L650). Am I correct that since All PoW chain forks use the same block fields format, so currently EVM doesn't handle it as well?